accepted/00000-jsonb-pillar-db-storage.md

```diff
@@ -0,0 +1,225 @@
+- Feature Name: Database backed salt pillar storage
+- Start Date: 2021-01-13
+- RFC PR: https://github.com/uyuni-project/uyuni-rfc/pull/51
+
+# Summary
+[summary]: #summary
+
+Use database as a storage for Uyuni salt pillars instead of plain files.
+
+# Motivation
+[motivation]: #motivation
+
+Pillars are used for passing reliable and trustworthy data to salt clients. Salt integration into Uyuni generates pillar files which are then loaded to the salt ecosystem by Uyuni external pillar.
+Advantage of this solution is rather easy implementation and quick visibility of the data when inspecting in filesystem, however there are disadvantages to consider. Data consistency [issue#5678](https://github.com/SUSE/spacewalk/issues/5678) and [issue#10679](https://github.com/SUSE/spacewalk/issues/10679), [filesystem performance](https://github.com/SUSE/spacewalk/pull/10391#discussion_r375147529), backup procedure must consider these files as well.
+
+This RFC proposes to instead of plain files, database based storage should be used for all pillars produced by Uyuni.
+
+This is expected to help with maintaining consistency, help with backup procedure and with interserver sync as well, potential performance improvement (TBD - filesystem access vs. database indexes).
```

Comment:
Do you intend to actually measure expected performance improvement over the base filesystem case as part of this RFC?

Reply:
Righ. The most I did is to do simple E2E performance test when I focused if there is noticable performance hit.
I reevaluated performance when I added one query option in form of postgres function.
Testing criteria are as follows:
* 1001 minions (one real, 1000 evil minions)
* 10 `pillar.items` repetitions for all minions
* 4000 rows in pillar database (1000 rows per organization, group, minion and global)
* regular Uyuni pillar files for one registered minion
* default Uyuni salt worker settings (8 worker threads)
* uptodate SUMA4.1 VM one machine (4vCPU, 4GiB RAM) and minions on real HW on different machine (8CPU, 24GiB RAM)

Command:

```bash
time bash -c 'for i in `seq 1 10`;do time salt "*" pillar.items >/dev/null;salt "*" test.ping > /dev/null;echo "== Round $i done"; done'       
```

And results are:
![performance-pillars](https://user-images.githubusercontent.com/1260795/114938853-e78f7980-9e3f-11eb-9ce6-cd7f866c8b50.png)

These results may be skewed by the fact that with the default settings I was occasionally hitting salt timeouts on some evil-minions. But these timeouts appeared on all of methods tested.

The best results were for single query and not using our external `suma_minion` pillar, next best was combined 1-query and `suma_minion` pillar, which is interesting that it was in average better then just using `suma_minion` pillar.

Reply:
Excellent, please transcribe some of this in the RFC's text - just as a reminder to posterity that proper benchmarking was conducted :-)